### PR TITLE
Add `USE_TSAN` option

### PR DESCRIPTION
**Pull request by malfet**
Sometimes it is important to run code with thread sanitizer.

Test Plan: CI

### CODE REVIEW

File: cmake/MiscCheck.cmake [hunk 1]
```diff
@@ -323,6 +323,15 @@ if (USE_ASAN)
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${CAFFE2_ASAN_FLAG}")
 endif()
 
+if (USE_TSAN)
+  set(CAFFE2_ASAN_FLAG "-fsanitize=thread -fPIE -pie")
```
Comment (orionr) [inline cmake/MiscCheck.cmake]: Nit: Do you want to have this be `CAFFE2_TSAN_FLAG` (instead of CAFFE2_ASAN_FLAG) for clarity and use it below?

**Review by orionr — APPROVED** [review]
Looks good! Just a nit.

File: cmake/MiscCheck.cmake (hunk 1 above)
Comment (malfet) [reply cmake/MiscCheck.cmake]: Updated.

**Comment by ezyang** [comment]
This is going to bitrot if it's not in CI ;) Also we're not TSAN clean today, AFAICT

**Pull request closed without merging**